### PR TITLE
fix: handle non-modulus-width elements in emulated IsZero

### LINKED ISSUE #1805: bug: emulated IsZero panics or returns wrong result for elements not at modulus width
## Description

`emulated.Field.IsZero` assumes the element it reduces always comes back with exactly the modulus limb count. Elements built by `Field.FromBits` (and other internal constructions) can legally carry a different number of limbs with zero overflow, and `Field.Reduce`'s fast path returns zero-overflow elements unchanged regardless of width. This breaks `IsZero` in two ways:

1. **Wider than the modulus → panic.** The `resP` comparison loop in `field_assert.go` is bounded by `len(ca.Limbs)` but indexes `p.Limbs[i]`:

```go
resP := f.api.IsZero(f.api.Sub(p.Limbs[0], ca.Limbs[0]))
for i := 1; i < len(ca.Limbs); i++ {
	resP = f.api.Mul(resP, f.api.IsZero(f.api.Sub(p.Limbs[i], ca.Limbs[i])))
}
```

Feeding e.g. a 512-bit `FromBits` result (8 limbs) into `IsZero` over a 4-limb modulus panics at circuit definition with `index out of range [4] with length 4`. This is reachable from std: `scalarMulFakeGLV` calls `IsZero(s)` on the incoming scalar, so any circuit passing a wide `FromBits` scalar (e.g. a SHA-512 output reduced mod L, as in Ed25519 verification) into `ScalarMul`/`MultiScalarMul` on a fakeGLV curve crashes at compile time.

2. **Correctness for non-modulus-width elements.** Even with the loop bounds fixed, the "reduced value is 0 or p" invariant only holds for values `< 2p`. A wide zero-overflow element can be any multiple of p (e.g. `2p`, `p²`), which would be misclassified as non-zero. Conversely, an element with *fewer* limbs than the modulus makes the loop compare only a prefix of `p.Limbs`, so a short element equal to the low limbs of p is misclassified as zero (false positive).

## Expected Behavior

`IsZero(a)` returns whether `a ≡ 0 (mod p)` for any width-constrained element, per its doc ("the method internally reduces the element").

## Actual Behavior

Panic for wider-than-modulus elements; wrong result possible for narrower-than-modulus elements.

## Steps to Reproduce

```go
type IsZeroWide[T emulated.FieldParams] struct {
	Bits []frontend.Variable // 2 * modulus bit length, e.g. a hash output
}

func (c *IsZeroWide[T]) Define(api frontend.API) error {
	f, err := emulated.NewField[T](api)
	if err != nil {
		return err
	}
	_ = f.IsZero(f.FromBits(c.Bits...))
	return nil
}

// frontend.Compile(ecc.BN254.ScalarField(), r1cs.NewBuilder,
//     &IsZeroWide[emulated.BN254Fr]{Bits: make([]frontend.Variable, 508)})
// => parse circuit: runtime error: index out of range [4] with length 4
```

Repro stack (via the std fakeGLV path):

```
runtime error: index out of range [4] with length 4
emulated.(*Field[...]).IsZero            std/math/emulated/field_assert.go
sw_emulated.(*Curve[...]).scalarMulFakeGLV   std/algebra/emulated/sw_emulated/point.go
sw_emulated.(*Curve[...]).jointScalarMulFakeGLV
sw_emulated.(*Curve[...]).MultiScalarMul
```

## Possible Fix

Force a real modular reduction in `IsZero` when the reduced element is wider than the modulus (the `Reduce` fast path skips zero-overflow elements regardless of limb count), and compare against all of `p.Limbs` (treating missing high limbs as zero) for narrower elements. PR incoming.

## Your Environment

- gnark `master` (48e4e4b67) and v0.13.0
- go 1.25.7, linux/amd64

### PR DESCRIPTION
**Pull request by 0x5ea000000**
# Description

`emulated.Field.IsZero` assumed the element it reduces always comes back with exactly the modulus limb count. Elements built by `Field.FromBits` can legally carry a different number of limbs with zero overflow, and `Field.Reduce`'s fast path returns zero-overflow elements unchanged regardless of width. This broke `IsZero` in two ways:

- **Wider than the modulus** (e.g. a 512-bit `FromBits` result over a 4-limb modulus): the limb-wise `ca == p` comparison indexed out of range and panicked at circuit definition (`index out of range [4] with length 4`). This is reachable from std: `scalarMulFakeGLV` calls `IsZero(s)` on the incoming scalar, so passing a wide `FromBits` scalar (e.g. a SHA-512 output for Ed25519-style verification) into `ScalarMul`/`MultiScalarMul` on a fakeGLV curve crashes at compile time. Additionally, such a value is not bounded by `2p`, so the 0-or-p classification would misclassify multiples of `p` (`2p`, `p²`, …) as non-zero even without the panic.
- **Narrower than the modulus**: the comparison loop only checked the given prefix of `p.Limbs`, returning a false positive (`IsZero == 1`) for a non-zero element equal to the low limbs of `p`.

The fix forces a full modular reduction to the modulus width when the reduced element is wider than the modulus (one `mulMod` by one, only on this path), and compares against every modulus limb, treating missing high limbs as zero, when it is narrower.

An alternative would be to tighten `Reduce`'s fast path itself (require `len(a.Limbs) <= NbLimbs` in addition to `overflow == 0`); that touches every consumer, so this PR keeps the change local to `IsZero`. Happy to rework in that direction if preferred.

Fixes #1805

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# How has this been tested?

- [x] New `TestIsZeroNonModulusWidth` (Goldilocks, BN254Fr, Mod1e512) covering: `2p` and `p²` as wide zero-overflow elements (`≡ 0 mod p`), `p² + 5` (non-zero), zero, and a single-limb element equal to the low limb of `p` (non-zero, previously a false positive). The wide cases panic and the narrow case fails before this fix.
- [x] `go test ./std/math/emulated/...` passes locally.

# How has this been benchmarked?

- [ ] Not benchmarked — the new `mulMod` reduction only triggers for elements wider than the modulus, which previously panicked; the width-matched path adds no constraints. The narrow-element path adds `len(p.Limbs) - len(ca.Limbs)` IsZero gates only for narrow elements.

# Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I did not modify files generated from templates
- [ ] `golangci-lint` does not output errors locally (not installed here; `gofmt` and `go vet` are clean)
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Correctness fix in a core emulated-field primitive used by scalar multiplication; wrong `IsZero` results can break circuit soundness. Change is localized to `IsZero` and covered by new tests.
> 
> **Overview**
> Fixes `Field.IsZero` for elements whose limb count differs from the modulus (e.g. `FromBits` results). `Reduce`'s zero-overflow fast path left these unchanged, so the 0-or-p check either **panicked** on wide elements or returned **false positives** on narrow ones.
> 
> **Wider than modulus:** force a full `mulMod` reduction first (with a small-field fold that avoids native wraparound). Also short-circuits constant elements at compile time.
> 
> **Narrower than modulus:** compare against every modulus limb, treating missing high limbs as zero.
> 
> Adds edge-case tests covering `2p`, `p²`, high-limb multiples, and single-limb prefixes of `p`, plus constant-wide elements.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 68a23d8b2b52048307d2f8e594cbbdda79956d25. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: std/math/emulated/field_assert.go [hunk 1]
```diff
@@ -194,9 +203,18 @@ func (f *Field[T]) IsZero(a *Element[T]) frontend.Variable {
 	// however, for checking if the element is p, we can not use the
 	// optimization as we may have underflows. So we have to check every limb
 	// individually.
+	// the element may have fewer limbs than the modulus (e.g. a short
+	// [Field.FromBits] result), in which case the missing high limbs are zero.
+	// We must still compare against every modulus limb — comparing only the
+	// prefix would return a false positive for an element equal to the low
+	// limbs of p.
```
[the remark this answers is not in the document]
Comment (0x5ea000000) [reply std/math/emulated/field_assert.go:245]: Confirmed — a fully-constant wide element that is 0 mod p folds through `mulMod`'s constant fast path to `newConstElement(0)`, which allocates zero limbs, and the limb-indexing below panicked. Fixed in d9ad695d1 by constant-folding `IsZero` at compile time for constant elements before any reduction (this also stops emitting constraints for constants generally), with a regression test (`TestIsZeroConstWide`) building wide constant elements for `2p` (expect 1) and `2p+5` (expect 0) — the former panicked with `index out of range [0] with length 0` before the fix.

File: std/math/emulated/field_assert.go [hunk 2]
```diff
@@ -164,6 +175,15 @@ func (f *Field[T]) IsZero(a *Element[T]) frontend.Variable {
...
 	p := f.Modulus()
+	if len(ca.Limbs) > len(p.Limbs) {
+		// [Field.Reduce] does not reduce zero-overflow elements even when they
+		// are wider than the modulus (e.g. a [Field.FromBits] result over more
+		// bits than the modulus width). Such a value is not bounded by 2p, so
+		// the 0-or-p check below would misclassify multiples of p — and the
+		// limb-wise comparison against the modulus would index out of range.
+		// Force a full modular reduction to the modulus width first.
+		ca = f.mulMod(ca, f.One(), 0, nil)
+	}
```
[the remark this answers is not in the document]
Comment (0x5ea000000) [reply std/math/emulated/field_assert.go:210]: Confirmed. Goldilocks over BN254 runs in small-field mode, and the wide-element reduction routed through `mulMod` → `toSingleLimbElement`, which recomposes limbs natively while keeping `overflow = 0`: at 3 limbs the understated overflow undersizes `smallMulMod`'s quotient range check (honest proving fails); at ≥ 4 limbs (≥ native bit length) the recomposition wraps, so `IsZero` answers for the wrong value. The previous `2·NbLimbs` tests (128 bits) coincidentally fit both bounds, which is why they passed.

Fixed in 68a23d8b2: in small-field mode the wide element is first folded into a single limb using mod-p-reduced coefficients (`Σ limbs[i]·(2^(w·i) mod p)` via native MulAcc) with an honestly-declared overflow (`overflow + pBits + ceil(log2(n))`), so the value is congruent mod p, fits the native field under the small-field-mode invariant (`2·modBits + 32 < native − 2`), and `smallMulMod`'s accounting is sized correctly. Tests now also cover `8·NbLimbs` (512 bits for Goldilocks — the native-wrap regime, mirroring the SHA-512 motivation) with multiples of p occupying the top limb; they fail without the fix and pass with it.

Note for maintainers: the underlying `toSingleLimbElement` behavior (native recomposition, unchanged overflow) is pre-existing and still reachable from other small-field `mulMod` callers fed multi-limb inputs (e.g. a wide `FromBits` result into `Mul`); this PR only hardens the `IsZero` path to keep its scope contained.
